### DISCUSSION TITLE
Deactivate docs nested routing globally?

### DISCUSSION THREAD

**Discussion by jliermann** [facebook/docusaurus#8414] [Q&A] [answered]
Dear friends,

We are looking for a way to have permalinks for the docs in our Docusaurus project because we are planning some changes in the project structure.
A straightforward way would have to be doc URLs in the format `domain.com/baseurl/docs/pagename`, agnostic of the actual location of the doc in the navigation structure. However, I cannot seem to find a way to globally deactivate nested routing for docs. Of course, one could also do that with slugs, but it would be very awkward to hardcode this in each doc's frontmatter.

Does such an option exist or are there any other thoughts on how to have persistent URLs for docs in Docusaurus?

Thanks a lot in advance and best regards,
Johannes

**Comment by slorber** [comment, marked as answer]
I am not sure to understand 100% of what you mean

You can get rid of the `/docs/` prefix with `routeBasePath: '/'`, and use `slug` to choose the doc URL you want.

We can't disable nested routing for docs because we are using shared layout elements between docs, context providers and other things.

At the top of a whole docs plugin instance is a provider that provides the version.

Similarly, we apply a shared layout on all docs, ensuring the docs sidebar remains mounted as you navigate across docs. If we had flat routes for docs, the full route would unmount/remount everytime and we'd lose our sidebar state, ie when you navigate, a category you expanded would collapse, and other weird side-effects that happen when local UI state gets reset.

Most React frameworks (Remix, Next.js...) move in the direction to use nested layouts for similar reason.

I think what you ask for is technically possible, but it would require a few important refactors and thinking, and is probably not worth it atm.

**Reply by slorber** [reply]
> A straightforward way would have to be doc URLs in the format domain.com/baseurl/docs/pagename, agnostic of the actual location of the doc in the navigation structure. However, I cannot seem to find a way to globally deactivate nested routing for docs.

We already support `domain.com/baseurl/docs/pagename`. The final URL does not depend on the "navigation structure" but rather the file-system structure. Using slug is the only way to decorrelate the final URL from the file-system location.

**Reply by jliermann** [reply]
Dear Sébastien, Thank you very much for the comprehensive answer.
What we are trying to do is make sure that links between docs don't break when the docs are reorganized in the directory structure. My first idea was to make sure that the URL of the docs would not change. But maybe this is not the best solution.
Is there possibly a way to make the links robust against changes in the file-system structure? Or are there other suggestions?

**Reply by slorber** [reply]
You can try using a slug
I also suggest linking to .md files (as we recommend in our doc).

We have the broken link checker to catch internal linking issues.

In general it's not a bad idea to assign a slug. Not all links are internal and users will bookmark some URLs and you will get external backlinks, so make sure to stabilize your URLs or use server redirects when appropriate.

**Comment by jliermann** [comment]
Dear Sébastien, sorry for getting back to this after a long time. We followed your suggestion and added slugs to all pages. To get only links like `/docs/page`, I wrote a small BASH script that, if applied to a file `page.mdx`, removes any slug and adds a slug `/page` to give the desired URL `/docs/page`. Maybe it's useful for other people.

```
#!/bin/bash

## Bash script to automatically add slug to frontmatter for permalinking
## Usage example: slugginator.sh *.mdx to "slug" all mdx files in present directory

# Regex for matching file name without number and ending

regex='[0-9]+\_([a-zA-Z0-9_-]*).mdx'

for i in "$@"
do
    # check if filename fits the pattern

    if [[ $i =~ $regex ]]; then

        #compose slug string from filename
        slug="slug: \"/${BASH_REMATCH[1]}\""

        echo "File $i: Slug is added to frontmatter:"
        echo "$slug"
        
        # Remove existing slug
        sed -i '/slug:.*/d' $i
        sed -i ':a;N;$!ba; s,---,'"$slug"'\n---,2' $i
    fi
 done
```
